### DISCUSSION TITLE
How to make authenticated calls when using NextJS with an external api.

### DISCUSSION THREAD

**Discussion by fitimbytyqi** [vercel/next.js#69451] [App Router]
Hello great community, I wanted to hear your thoughts on this topic.

In my projects, I always use NextJS alongside an external API. When I login, I like to save the JWT token in server-side cookies which is done on the nextJS server side, the external server only returns me the JWT and other informations.

When I want to fetch or mutate data from server components I can access that token from next/header cookies() function to forward it to that external API via Authorization header of fetch, however I'm confused when it comes to client side components.

Currently I am using TanStack Query but my question is how do I access that JWT token which lives in the server cookie?

Ofc one of the solutions is to overload the developer experience and use something like TRPC and use NextJS as BFF.

Any suggestions or anything that I'm not aware of is welcomed, thank you.

**Comment by Juneezee** [comment]
> I like to save the JWT token in server-side cookies

Not sure if I understood you correcly but cookies are actually stored on the browser, not the server.

Open up the browser developer tool and you can see your cookie
![image](https://github.com/user-attachments/assets/42aff1ab-f4d3-4837-ac85-1d41064c1a05)

> but my question is how do I access that JWT token 

You can either

1. Pass the cookie from Server Component to Client Component through props
2. Read the cookie on Client Component through `document.cookie` (https://developer.mozilla.org/en-US/docs/Web/API/Document/cookie)

**Reply by fitimbytyqi** [reply]
The cookies I'm talking about are secured with security flags which make it impossible to access via document.cookie

**Reply by Juneezee** [reply]
In that case Option 1 seems to be the only possible way to do it securely.

**Reply by fitimbytyqi** [reply]
Another solution would be to use NextJS as BFF, meaning that every fetch or mutation must go through the next server, there I can get the token and do whatever with it and that's what I am doing but I was wondering if there is anything else I'm missing.

**Comment by BesoKakulia** [comment]
If you're using the same domain for both your external server and your Next.js app, ex: `api.somedomain.com` for the API and `app.somedomain.com` for the app, you can forward cookies from your Next.js requests to the client. The browser will then automatically set these cookies and include them in any requests sent to your API server from the client.

**Reply by fitimbytyqi** [reply]
@BesoKakulia Yeah I know that but in my case it's different, the api has another domain.

**Reply by BesoKakulia** [reply]
Ah, ok I have the same problem then, I look for answer here as well 😅
Though, seems like 3rd party cookies won't be allowed soon.
`Third-party cookie will be blocked in future Chrome versions as part of Privacy Sandbox.`

**Reply by fitimbytyqi** [reply]
HTTP-only secure cookies won't be affected by this feature AFAIK

**Reply by BesoKakulia** [reply]
Ok, I think I got it, works for my problem.
I can make request to the external api server from both Client and Server components.
My cookies are httpOnly with same origin flag.

These 2 posts helped me https://maxschmitt.me/posts/next-js-http-only-cookie-auth-tokens https://maxschmitt.me/posts/next-js-api-proxy

**Comment by sundargautam18** [comment]
You can create your own: Here is the better and simple solution to implement this : https://medium.com/@sundargautam2022/implementing-refresh-token-with-nextjs-15-using-app-router-with-cross-api-different-api-5682f83f9802
